### ISSUE TITLE
preCICE xml files are generally not namespace-well-formed

### ISSUE THREAD

**Issue by ajaust** [precice/precice#928]
**Describe your setup**

Operating system (e.g. Linux distribution and version): Ubuntu 18.04
preCICE Version: Should be independent of preCICE version

**Describe the problem**
The XML files are hard to parse in common XML parsers since no namespaces are defined.

The XML files used to configure preCICE use colon notation as in `data:scalar`. This can be interpreted as XML namespace. However, it is not common to provide XML namespaces in preCICE. The tutorials are missing them for example and also the [documentation](https://www.precice.org/configuration-introduction.html) starts with a simple `<precice-configuration>` tag not defining namespaces. In order to parse such a file one has to set non-standard parser options. This could be avoided if the XML would provide XML namespaces.

**Step To Reproduce**
1. Download the attachement
2. Make sure you have the needed Python modules
3. Run the `parse-xml.py` file included and 3 of 4 parser should fail. The fourth case does not used the standard options of the parser.
4. Change the parsed file in `parse-xml.py` to `precice-config-1d-fsi-python-fixed.xml` to see how it could work with minimal changes to the XML.

**Expected behaviour**
The XML can be parsed with common XML parsers without setting non-standard options.

**Additional context**

Adding namespace declarations to a configuration file did not create any problems for preCICE when I tested it.

More information on stackoverflow and the w3 specs:
- https://stackoverflow.com/questions/134494/is-there-any-difference-between-valid-xml-and-well-formed-xml/25830482#25830482
- https://www.w3.org/TR/REC-xml/#sec-common-syn

I have attached some example code that tries to parse the XML file in Python. It also includes a potentially fixed version of the preCICE configuration which defines namespaces.
[parse-precice-xml.tar.gz](https://github.com/precice/precice/files/5786930/parse-precice-xml.tar.gz)

**Comment by BenjaminRodenberg** [comment]
Hi,

we already discussed this issue previously in https://github.com/precice/precice/issues/127 and decided to make it a `wontfix`. I also stumbled across this issue when trying to use a python-based xml parser. Can you share your opinion on this topic under #127 ? My knowledge of xml is quite limited so it's absolutely possible that I missed some important points in #127 and you can give us more insight.

I would suggest to (at least) keep #928 open until we get to a conclusion on this topic.

Cheers!

**Comment by ajaust** [comment]
Ah. I missed the old issue when opening this one. I guess it is fine to not fix it. It is your software package and it is nice that it is using a XML structure. Maybe it would be worth to mention in the documentation that the namespaces are missing and an example what to add if one wants to parse such an XML file? It would also be possible to add the namespaces when updating the tutorials without requiring them in the parser and thus keeping the backwards compatibility. preCICE parses well-formed XML without any issues. I have a XML generator that defines namespaces and it has not been a problem.

**Comment by ajaust** [comment]
After some more time thinking about this topic I would have the following suggestions:

- Add a note regarding what to do if some tool complains about the XML format (i.e. add namespaces)
- Adding the namespaces to the tutorials, documentation etc. (over some time) to slowly teach users a new best practice
- Do not enforce the use of namespaces (for now) to keep compatibility. 

Adding the namespaces should not be much more than extending the first line containing `precice-configuration`. When I downloaded a XML configuration file recently my OS considered a browser to be the "natural" choice to view XML files. However, this gave me just some errors:

```
This page contains the following errors:
error on line 11 at column 34: Namespace prefix data on vector is not defined
error on line 12 at column 46: Namespace prefix data on vector is not defined
error on line 33 at column 123: Namespace prefix mapping on nearest-neighbor is not defined
error on line 47 at column 47: Namespace prefix m2n on sockets is not defined
error on line 49 at column 39: Namespace prefix coupling-scheme on serial-implicit is not defined
error on line 59 at column 34: Namespace prefix acceleration on IQN-IMVJ is not defined
```
I could see that these errors, that do not affect preCICE, might still confuse some users at some point.

If you would decide to enforce the namespace at some time, you could do so by checking for the lack of namespaces within for preCICE for some releases and issuing a warning. In that case people would have to time to get adapted to this.

**Comment by fsimonis** [comment]
This is a problematic matter, as we "misuse" the namespace feature to make the XML more readable. Defining the namespaces makes the less readable again.

Instead of adding namespaces, we would rather use a correct way of solving this.
Example solutions could be attributes `<m2n type="sockets" />` or compound names  `<m2n-sockets ... />`. Both cases requires to rewrite the complete configuration back-end and breaks all existing configuration files.

If we decide to break configuration files, we would go the extra step and clean up the complete configuration, possibly even choosing a more suitable format. Example could be #168. 

A current workaround for custom scripts is to use an XML parser that does not enforce correct namespaces. This is why I used `lxml` in the [config-visualizer](https://github.com/precice/config-visualizer/blob/79d7c1c3f04bf1f23b2ea463eb7d1445ab720afc/bin/precice-config-visualizer#L3).

So, we are moving this to version 3.
